pkg/bucket/bucket.go

```diff
@@ -29,33 +28,56 @@ const (
 )
 
 type Bucket struct {
-	buf []byte
-	src *[]byte
-
+	slots              [][]byte
 	init               bool
 	resyncOnNextPacket bool
 	step               int
 	headSN             uint16
+	initCapacity       int
 	maxSteps           int
 }
 
-func NewBucket(buf *[]byte) *Bucket {
+func NewBucket(capacity int) *Bucket {
 	b := &Bucket{
-		src:      buf,
-		buf:      *buf,
-		maxSteps: int(math.Floor(float64(len(*buf)) / float64(MaxPktSize))),
+		initCapacity: capacity,
+		maxSteps:     capacity,
+		slots:        make([][]byte, capacity),
+	}
+
+	pktSize := MaxPktSize + pktSizeHeader
+	buf := make([]byte, pktSize*capacity)
+	for i := 0; i < capacity; i++ {
+		b.slots[i] = buf[i*pktSize : (i+1)*pktSize]
 	}
 
 	b.invalidate(0, b.maxSteps)
 	return b
 }
 
-func (b *Bucket) ResyncOnNextPacket() {
-	b.resyncOnNextPacket = true
+// Grow increases the capacity of the bucket by adding initial capacity to the buffer
+func (b *Bucket) Grow() int {
+	pktSize := MaxPktSize + pktSizeHeader
+	buf := make([]byte, pktSize*b.initCapacity)
+	newSlots := make([][]byte, b.initCapacity)
+	for i := 0; i < b.initCapacity; i++ {
+		newSlots[i] = buf[i*pktSize : (i+1)*pktSize]
+		binary.BigEndian.PutUint16(newSlots[i], invalidPktSize)
```

Comment:
Do this in the above constructor. Or just make a function to pass in the buf, slots and do it there?